File: Telephone/SystemMediaPlayer.m
//
//  SystemMediaPlayer.m
//  Telephone
//
//  Private MediaRemote bridge for this personal macOS fork.
//

#import "SystemMediaPlayer.h"

#import <dlfcn.h>

typedef void (*MRGetNowPlayingApplicationIsPlayingFunction)(
    dispatch_queue_t queue,
    void (^completion)(Boolean isPlaying)
);
typedef Boolean (*MRSendCommandFunction)(NSInteger command, NSDictionary * _Nullable userInfo);

static NSString * const kMediaRemotePath =
    @"/System/Library/PrivateFrameworks/MediaRemote.framework/MediaRemote";
static const NSInteger kMediaRemotePlayCommand = 0;
static const NSInteger kMediaRemotePauseCommand = 1;

@interface SystemMediaPlayer ()

@property(nonatomic) void *mediaRemoteHandle;
@property(nonatomic) MRGetNowPlayingApplicationIsPlayingFunction getIsPlaying;
@property(nonatomic) MRSendCommandFunction sendCommand;
@property(nonatomic) BOOL pauseRequested;
@property(nonatomic) BOOL didPause;
@property(nonatomic) NSUInteger requestGeneration;

@end

@implementation SystemMediaPlayer

- (instancetype)init {
    self = [super init];
    if (self == nil) {
        return nil;
    }

    _mediaRemoteHandle = dlopen(kMediaRemotePath.fileSystemRepresentation, RTLD_LAZY | RTLD_LOCAL);
    if (_mediaRemoteHandle == NULL) {
        NSLog(@"System media control unavailable: could not load MediaRemote");
        return self;
    }

    _getIsPlaying = (MRGetNowPlayingApplicationIsPlayingFunction)
        dlsym(_mediaRemoteHandle, "MRMediaRemoteGetNowPlayingApplicationIsPlaying");
    _sendCommand = (MRSendCommandFunction)
        dlsym(_mediaRemoteHandle, "MRMediaRemoteSendCommand");

    if (_getIsPlaying == NULL || _sendCommand == NULL) {
        NSLog(@"System media control unavailable: required MediaRemote symbols are missing");
    }

    return self;
}

- (void)dealloc {
    if (_mediaRemoteHandle != NULL) {
        dlclose(_mediaRemoteHandle);
    }
}

- (void)pause {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.pauseRequested = YES;
        NSUInteger generation = ++self.requestGeneration;

        if (self.getIsPlaying == NULL || self.sendCommand == NULL) {
            return;
        }

        self.getIsPlaying(dispatch_get_main_queue(), ^(Boolean isPlaying) {
            if (!self.pauseRequested ||
                generation != self.requestGeneration ||
                !isPlaying) {
                return;
            }

            if (self.sendCommand(kMediaRemotePauseCommand, nil)) {
                self.didPause = YES;
            }
        });
    });
}

- (void)resume {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.pauseRequested = NO;
        ++self.requestGeneration;

        if (!self.didPause) {
            return;
        }

        self.didPause = NO;
        if (self.sendCommand != NULL) {
            self.sendCommand(kMediaRemotePlayCommand, nil);
        }
    });
}

@end

File: Telephone/SystemMediaPlayer.swift
//
//  SystemMediaPlayer.swift
//  Telephone
//
//  Private MediaRemote bridge for this personal macOS fork.
//

import Darwin
import Foundation
import UseCases

final class SystemMediaPlayer: NSObject, MusicPlayer {
    private typealias IsPlayingCompletion =
        @convention(block) (UInt8) -> Void
    private typealias GetIsPlaying =
        @convention(c) (DispatchQueue, IsPlayingCompletion) -> Void
    private typealias SendCommand =
        @convention(c) (Int, NSDictionary?) -> UInt8

    private static let mediaRemotePath =
        "/System/Library/PrivateFrameworks/MediaRemote.framework/MediaRemote"
    private static let playCommand = 0
    private static let pauseCommand = 1

    private var mediaRemoteHandle: UnsafeMutableRawPointer?
    private var getIsPlaying: GetIsPlaying?
    private var sendCommand: SendCommand?

    private var pauseRequested = false
    private var didPause = false
    private var requestGeneration = 0

    override init() {
        super.init()

        mediaRemoteHandle = dlopen(
            Self.mediaRemotePath,
            RTLD_LAZY | RTLD_LOCAL
        )

        guard let mediaRemoteHandle else {
            NSLog(
                "System media control unavailable: could not load MediaRemote"
            )
            return
        }

        if let symbol = dlsym(
            mediaRemoteHandle,
            "MRMediaRemoteGetNowPlayingApplicationIsPlaying"
        ) {
            getIsPlaying = unsafeBitCast(symbol, to: GetIsPlaying.self)
        }

        if let symbol = dlsym(
            mediaRemoteHandle,
            "MRMediaRemoteSendCommand"
        ) {
            sendCommand = unsafeBitCast(symbol, to: SendCommand.self)
        }

        if getIsPlaying == nil || sendCommand == nil {
            NSLog(
                "System media control unavailable: required MediaRemote symbols are missing"
            )
        }
    }

    deinit {
        if let mediaRemoteHandle {
            dlclose(mediaRemoteHandle)
        }
    }

    func pause() {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }

            pauseRequested = true
            requestGeneration &+= 1
            let generation = requestGeneration

            guard
                let getIsPlaying,
                let sendCommand
            else {
                return
            }

            getIsPlaying(.main) { [weak self] isPlaying in
                guard
                    let self,
                    pauseRequested,
                    generation == requestGeneration,
                    isPlaying != 0
                else {
                    return
                }

                if sendCommand(Self.pauseCommand, nil) != 0 {
                    didPause = true
                }
            }
        }
    }

    func resume() {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }

            pauseRequested = false
            requestGeneration &+= 1

            guard didPause else { return }
            didPause = false

            _ = sendCommand?(Self.playCommand, nil)
        }
    }
}
